Objective-C/CBLDatabase.mm

```diff
@@ -790,13 +812,83 @@ - (void) freeC4DB {
 }
 
 
-#pragma mark - DOCUMENT SAVE AND CONFLICT RESOLUTION
+#pragma mark - DOCUMENT SAVE AND CONFLICT HANDLING
+
+
+- (BOOL) saveDocument: (CBLDocument*)document
+   concurrencyControl: (CBLConcurrencyControl)concurrencyControl
+           asDeletion: (BOOL)deletion
+                error: (NSError**)outError
+{
+    if (deletion && !document.revID)
+        return createError(CBLErrorInvalidParameter,
+                           @"Document doesn't exist in the database.", outError);
+    
+    CBL_LOCK(self) {
+        if (![self prepareDocument: document error: outError])
+            return NO;
+        
+        C4Document* curDoc = nil;
+        C4Document* newDoc = nil;
+        @try {
+            // Begin a db transaction:
+            C4Transaction transaction(_c4db);
+            if (!transaction.begin())
+                return convertError(transaction.error(), outError);
+            
+            if (![self saveDocument: document into: &newDoc
+                   withBaseDocument: nil asDeletion: deletion error: outError])
+                return NO;
+            
+            if (!newDoc) {
+                // Handle conflict:
+                if (concurrencyControl == kCBLConcurrencyControlFailOnConflict)
+                    return createError(CBLErrorConflict, outError);
+                
+                C4Error err;
+                CBLStringBytes bDocID(document.id);
+                curDoc = c4doc_get(_c4db, bDocID, true, &err);
+                
+                // If deletion and the current doc has already been deleted
+                // or doesn't exist:
+                if (deletion) {
+                    if (!curDoc) {
+                        if (err.code == kC4ErrorNotFound)
+                            return YES;
+                        return convertError(err, outError);
+                    } else if ((curDoc->flags & kDocDeleted) != 0) {
+                        [document replaceC4Doc: [CBLC4Document document: curDoc]];
+                        curDoc = nil;
+                        return YES;
+                    }
+                }
+                
+                // Save changes on the current branch:
+                assert(curDoc);
```

Comment:
This doesn't take into account an error getting curDoc (unless `deletion` is true.) For example, saving a doc when another thread has purged it. Or some I/O error on the database, perhaps.

Reply:
That's true.